### ISSUE TITLE
savegame system

### ISSUE THREAD

**Issue by herobank110** [herobank110/factorygame#5]
SaveGame objects with serialisable properties that can be saved and loaded into save slots

**Comment by gudeliauskaspam** [comment]
I'm interested in trying to do this and if everything goes well, also in opening a PR during Hacktoberfest. :)

**Comment by herobank110** [comment]
Hi, thanks for your interest!

Let me elaborate in the issue. It is for a data saving and loading mechanism for games, such as the last checkpoint, player name, high score, etc.

I was thinking of 2 new functions:
saving function - _input:_ dictionary to save, slot name, _output:_ whether the save was successful
loading function - _input:_ slot name, output: dictionary of saved values, or None if the slot couldn't be loaded
If you have other suggestions please let me know.

A good place for this would be as static methods of GameplayUtilities, under `factorygame/utils/gameplay.py`

My initial idea was to have a SaveGame class to add fields to in a derived class, so you could have autocomplete on the loaded save game object, but this should be a separate issue.

**Comment by gudeliauskaspam** [comment]
I was thinking about the same approach of two new functions to save and load.
I really want to work on this challenge, I found your project very interesting =)

**Comment by herobank110** [comment]
Thanks. I assigned you to the issue!

**Comment by gudeliauskaspam** [comment]
hello, sorry for the lack of news.
this week I start working on this problem
:)

**Comment by gudeliauskaspam** [comment]
can i use any library?

**Comment by herobank110** [comment]
what library were you thinking of using?

**Comment by gudeliauskaspam** [comment]
some library like Pickle

**Comment by herobank110** [comment]
Certainly, anything in the standard library is fine. Other libraries might be difficult to link as the package doesn't currently have a setup.py.
If you do use pickle please consider Protocol version 3 which is compatible with python 3.2, the oldest version of python currently supported by this project.

**Comment by p4mela-g** [comment]
Do you think about supporting python 3.7+ in this application?

**Comment by p4mela-g** [comment]
This is what I'm doing:

In the save / create method, I check if the folder containing the savegame exists, if it does not exist, it is created. In order not to have problems with overwriting files, I think of using a standard name like "savegame_nowtimestamp.dat".
In the load method, I think about checking if the file you want exists, and if it exists, I load the data.
In both cases, a message will be sent to the user if the action was successful or not.

If you are dissatisfied with any of the steps above, tell me and I will do it the way you think is ideal.
 :smile: 

I'm having a little trouble finding where to collect the data to be saved on the savegame.

**Comment by herobank110** [comment]
Great work! Sounds like you have a lot of functionality.

I have a few ideas for you:
- Can you use a .sav instead of .dat file extension?
   dat might be used for resource bundles.
- Can you remove the timestamp from the file name?
   It is fine to overwrite a save game in a slot when you save and I don't think automatic backup is necessary. But it may be useful in some cases like quick-save or checkpoint save slots so please feel free to pursue this in a new issue!
- Can you put the slot name in the file name?
   Not sure if you are already doing this but 'savegame_timestamp.dat' made me uncertain. Ideally something like this: '{slot_name}.sav'.
- Can you add a function to check if a save game exists? Eg, does_save_game_exist(slot_name)
   This would allow devs to implement a warning prompt before overwriting a slot. It wasn't part of the issue originally so maybe we can leave it to a new issue?

The data that should be saved you have two options:
- a dictionary passed to the save function. The data would already be in key-value format that could be saved.
- an instance of a specific class. Pick any members not beginning with underscore that can be serialized. The difficulty with this is class structure may change and python's dynamic types. Solving these could definitely be its own issue.
So probably taking a dictionary is the best option?

I definitely want to support python 3.7+. Someday I will drop 3.2 support and make 3.6 the oldest supported.

**Comment by p4mela-g** [comment]
Sorry! I had some issues with my OS and I ended up losing progress before I could commit.
I'll start over.
